src/follower.ts

```diff
@@ -0,0 +1,100 @@
+import Resource from './resource';
+import { LinkVariables } from './types';
+
+/**
+ * The Follower class is what's being returned from follow() functions.
+ *
+ * It's 'PromiseLike', which means you can treat it like a Promise, and it
+ * can be awaited. When used as a Promise, it resolves to the Resource object
+ * that was followed.
+ *
+ * In addition to being a Promise<Resource> stand-in, it also exposes other
+ * functions, namely:
+ *
+ * * `follow()` to allow a user to chain several follow() functions to do
+ *   several 'hops' all at once.
+ * * `followAll()`, allowing a user to call `followAll()` at the end of a
+ *   chain.
+ */
+export default class Follower implements PromiseLike<Resource> {
+
+  private resource: Resource | Promise<Resource>;
+  private rel: string;
+  private variables?: LinkVariables;
+
+  constructor(resource: Resource | Promise<Resource>, rel: string, variables?: LinkVariables) {
+
+    this.resource = resource;
+    this.rel = rel;
+    this.variables = variables;
+
+  }
+
+  /**
+   * This 'then' function behaves like a Promise then() function.
+   */
+  then<TResult1 = Resource<any>, TResult2 = never>(onfulfilled?: ((value: Resource<any>) => TResult1 | PromiseLike<TResult1>) | null | undefined, onrejected?: ((reason: Error) => TResult2 | PromiseLike<TResult2>) | null | undefined): Promise<TResult1 | TResult2> {
```

Comment:
Would you mind defining the input and output types separately from the function? It would make it easier to read.